File: services/speaker-coordination-service/src/main/java/ch/batbern/speakers/service/SpeakerStatusService.java
package ch.batbern.speakers.service;

import ch.batbern.shared.exception.NotFoundException;
import ch.batbern.shared.types.SpeakerWorkflowState;
import ch.batbern.speakers.domain.SpeakerStatusHistory;
import ch.batbern.speakers.dto.SpeakerStatusResponse;
import ch.batbern.speakers.dto.StatusHistoryItem;
import ch.batbern.speakers.dto.StatusSummaryResponse;
import ch.batbern.speakers.dto.UpdateStatusRequest;
import ch.batbern.speakers.repository.SpeakerStatusHistoryRepository;
import ch.batbern.speakers.validator.StatusTransitionValidator;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import static ch.batbern.speakers.config.CacheConfig.STATUS_HISTORY_CACHE;
import static ch.batbern.speakers.config.CacheConfig.STATUS_SUMMARY_CACHE;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Service for managing speaker status transitions
 * Story 5.4: Speaker Status Management - Task 5 (GREEN Phase)
 *
 * Handles:
 * - Status updates with validation (AC1-2)
 * - Status history tracking (AC3-4)
 * - Status summary calculation with acceptance rate (AC5-6)
 * - Overflow detection (AC13)
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class SpeakerStatusService {

    private final SpeakerStatusHistoryRepository repository;
    private final StatusTransitionValidator validator;

    // TODO: These should come from Event configuration - hardcoded for now
    private static final int DEFAULT_MIN_SPEAKERS = 6;
    private static final int DEFAULT_MAX_SPEAKERS = 8;

    /**
     * Update speaker status with validation
     * Story 5.4 AC1-2: Manual status updates with workflow validation
     * Cache eviction: Invalidates both status summary and history caches for the event
     *
     * @param eventCode Event code
     * @param speakerId Speaker pool ID
     * @param organizerUsername Username of organizer making the change
     * @param request Update request with new status and optional reason
     * @return Status change response
     */
    @CacheEvict(value = {STATUS_SUMMARY_CACHE, STATUS_HISTORY_CACHE}, key = "#eventCode")
    public SpeakerStatusResponse updateStatus(
        String eventCode,
        UUID speakerId,
        String organizerUsername,
        UpdateStatusRequest request
    ) {
        log.info("Updating speaker {} status to {} for event {} by {}",
            speakerId, request.getNewStatus(), eventCode, organizerUsername);

        // Get current status from latest history record
        List<SpeakerStatusHistory> history = repository.findBySpeakerPoolIdOrderByChangedAtDesc(speakerId);
        SpeakerWorkflowState currentStatus = history.isEmpty()
            ? SpeakerWorkflowState.IDENTIFIED
            : history.get(0).getNewStatus();

        // Validate state transition - Story 5.4 AC12
        validator.validateTransition(currentStatus, request.getNewStatus());

        // Create history record - Story 5.4 AC3-4
        SpeakerStatusHistory historyRecord = new SpeakerStatusHistory();
        historyRecord.setSpeakerPoolId(speakerId);
        historyRecord.setSessionId(UUID.randomUUID()); // TODO: Get actual session ID
        historyRecord.setEventCode(eventCode);
        historyRecord.setPreviousStatus(currentStatus);
        historyRecord.setNewStatus(request.getNewStatus());
        historyRecord.setChangedByUsername(organizerUsername);
        historyRecord.setChangeReason(request.getReason());
        historyRecord.setChangedAt(Instant.now());

        SpeakerStatusHistory saved = repository.save(historyRecord);

        // Build response
        return mapToResponse(saved);
    }

    /**
     * Get status change history for a speaker
     * Story 5.4 AC15: Query status history
     * Cached for 60 seconds per event (Story 5.4 cache requirement)
     *
     * @param eventCode Event code
     * @param speakerId Speaker pool ID
     * @return List of status changes ordered by time descending
     */
    @Transactional(readOnly = true)
    @Cacheable(value = STATUS_HISTORY_CACHE, key = "#eventCode")
    public List<StatusHistoryItem> getStatusHistory(String eventCode, UUID speakerId) {
        log.debug("Fetching status history for speaker {} in event {}", speakerId, eventCode);

        List<SpeakerStatusHistory> history = repository.findBySpeakerPoolIdOrderByChangedAtDesc(speakerId);

        if (history.isEmpty()) {
            throw new NotFoundException("No status history found for speaker: " + speakerId);
        }

        return history.stream()
            .map(this::mapToHistoryItem)
            .collect(Collectors.toList());
    }

    /**
     * Get status summary with counts and acceptance rate
     * Story 5.4 AC5-6: Status dashboard with acceptance rate
     * Story 5.4 AC13: Overflow detection
     * Cached for 60 seconds per event (Story 5.4 cache requirement - TTL: 60s)
     *
     * @param eventCode Event code
     * @return Status summary with counts, rate, and overflow detection
     */
    @Transactional(readOnly = true)
    @Cacheable(value = STATUS_SUMMARY_CACHE, key = "#eventCode")
    public StatusSummaryResponse getStatusSummary(String eventCode) {
        log.debug("Calculating status summary for event {}", eventCode);

        // Get all speakers for the event (latest status for each)
        // Note: This is simplified - in production, we'd query speaker_pool table
        // For now, we'll get unique speakers from history and their latest status

        List<SpeakerStatusHistory> allHistory = repository.findAll();

        // Group by speaker and get latest status
        Map<UUID, SpeakerWorkflowState> speakerStatuses = allHistory.stream()
            .filter(h -> h.getEventCode().equals(eventCode))
            .collect(Collectors.groupingBy(
                SpeakerStatusHistory::getSpeakerPoolId,
                Collectors.collectingAndThen(
                    Collectors.maxBy((h1, h2) -> h1.getChangedAt().compareTo(h2.getChangedAt())),
                    opt -> opt.map(SpeakerStatusHistory::getNewStatus).orElse(SpeakerWorkflowState.IDENTIFIED)
                )
            ));

        // Calculate counts - Story 5.4 AC5
        Map<SpeakerWorkflowState, Long> statusCounts = speakerStatuses.values().stream()
            .collect(Collectors.groupingBy(s -> s, Collectors.counting()));

        long totalSpeakers = speakerStatuses.size();
        long acceptedCount = statusCounts.getOrDefault(SpeakerWorkflowState.ACCEPTED, 0L);
        long declinedCount = statusCounts.getOrDefault(SpeakerWorkflowState.DECLINED, 0L);
        long pendingCount = totalSpeakers - acceptedCount - declinedCount;

        // Calculate acceptance rate - Story 5.4 AC6
        double acceptanceRate = (totalSpeakers > 0)
            ? (acceptedCount * 100.0 / totalSpeakers)
            : 0.0;

        // Check thresholds and overflow - Story 5.4 AC13
        boolean thresholdMet = acceptedCount >= DEFAULT_MIN_SPEAKERS;
        boolean overflowDetected = acceptedCount > DEFAULT_MAX_SPEAKERS;

        StatusSummaryResponse response = new StatusSummaryResponse();
        response.setEventCode(eventCode);
        response.setStatusCounts(statusCounts);
        response.setTotalSpeakers(totalSpeakers);
        response.setAcceptedCount(acceptedCount);
        response.setDeclinedCount(declinedCount);
        response.setPendingCount(pendingCount);
        response.setAcceptanceRate(Math.round(acceptanceRate * 100.0) / 100.0);
        response.setMinSlotsRequired(DEFAULT_MIN_SPEAKERS);
        response.setMaxSlotsAllowed(DEFAULT_MAX_SPEAKERS);
        response.setThresholdMet(thresholdMet);
        response.setOverflowDetected(overflowDetected);

        log.info("Status summary for {}: {} total, {} accepted ({}%), overflow: {}",
            eventCode, totalSpeakers, acceptedCount, response.getAcceptanceRate(), overflowDetected);

        return response;
    }

    /**
     * Map entity to response DTO
     */
    private SpeakerStatusResponse mapToResponse(SpeakerStatusHistory history) {
        SpeakerStatusResponse response = new SpeakerStatusResponse();
        response.setSpeakerId(history.getSpeakerPoolId());
        response.setEventCode(history.getEventCode());
        response.setCurrentStatus(history.getNewStatus());
        response.setPreviousStatus(history.getPreviousStatus());
        response.setChangedByUsername(history.getChangedByUsername());
        response.setChangeReason(history.getChangeReason());
        response.setChangedAt(history.getChangedAt());
        return response;
    }

    /**
     * Map entity to history item DTO
     */
    private StatusHistoryItem mapToHistoryItem(SpeakerStatusHistory history) {
        StatusHistoryItem item = new StatusHistoryItem();
        item.setId(history.getId());
        item.setPreviousStatus(history.getPreviousStatus());
        item.setNewStatus(history.getNewStatus());
        item.setChangedByUsername(history.getChangedByUsername());
        item.setChangeReason(history.getChangeReason());
        item.setChangedAt(history.getChangedAt());
        return item;
    }
}

File: web-frontend/src/services/speakerStatusService.ts
/**
 * Speaker Status Service (Story 5.4)
 *
 * HTTP client for Speaker Status Management APIs
 * Features:
 * - Update speaker status with workflow validation
 * - Get speaker status history with audit trail
 * - Get status dashboard summary with acceptance metrics
 * - JWT authentication via interceptors
 * - Error handling with correlation IDs
 */

import apiClient from '@/services/api/apiClient';
import type { components } from '@/types/generated/speakers-api.types';

// Type aliases for generated types
type UpdateStatusRequest = components['schemas']['UpdateStatusRequest'];
type SpeakerStatusResponse = components['schemas']['SpeakerStatusResponse'];
type StatusHistoryItem = components['schemas']['StatusHistoryItem'];
type StatusSummaryResponse = components['schemas']['StatusSummaryResponse'];
type SpeakerWorkflowState = components['schemas']['SpeakerWorkflowState'];

// API base path
const EVENTS_API_PATH = '/events';

/**
 * Speaker Status Service
 *
 * Handles all HTTP requests for speaker status management
 */
class SpeakerStatusService {
  /**
   * Update speaker status (AC1-2, AC10-12)
   *
   * Valid state transitions:
   * - OPEN → CONTACTED
   * - CONTACTED → READY, DECLINED
   * - READY → ACCEPTED, DECLINED
   * - ACCEPTED → SLOT_ASSIGNED (cannot go back to DECLINED)
   * - DECLINED is terminal (cannot transition out)
   *
   * @param eventCode Event code (e.g., "BATbern56")
   * @param speakerId Speaker pool UUID
   * @param newStatus New workflow state
   * @param reason Optional reason for status change (max 2000 characters)
   * @returns Updated speaker status with change details
   * @throws Error if validation fails (400), speaker not found (404), or invalid transition (422)
   */
  async updateStatus(
    eventCode: string,
    speakerId: string,
    newStatus: SpeakerWorkflowState,
    reason?: string
  ): Promise<SpeakerStatusResponse> {
    const request: UpdateStatusRequest = {
      newStatus,
      ...(reason && { reason }),
    };

    const response = await apiClient.put<SpeakerStatusResponse>(
      `${EVENTS_API_PATH}/${eventCode}/speakers/${speakerId}/status`,
      request
    );

    return response.data;
  }

  /**
   * Get speaker status history (AC3-4, AC15)
   *
   * Retrieves complete status change history with timestamps, organizers, and reasons
   *
   * @param eventCode Event code (e.g., "BATbern56")
   * @param speakerId Speaker pool UUID
   * @returns Array of status history items ordered by changedAt (newest first)
   * @throws Error if speaker not found (404) or unauthorized (401, 403)
   */
  async getStatusHistory(eventCode: string, speakerId: string): Promise<StatusHistoryItem[]> {
    const response = await apiClient.get<StatusHistoryItem[]>(
      `${EVENTS_API_PATH}/${eventCode}/speakers/${speakerId}/status/history`
    );

    return response.data;
  }

  /**
   * Get status dashboard summary (AC5-6, AC13)
   *
   * Returns aggregated speaker status metrics including:
   * - Status counts by workflow state
   * - Acceptance rate percentage
   * - Threshold status (min/max slots)
   * - Overflow detection
   *
   * @param eventCode Event code (e.g., "BATbern56")
   * @returns Status summary with counts, rates, and threshold indicators
   * @throws Error if event not found (404) or unauthorized (401, 403)
   */
  async getStatusSummary(eventCode: string): Promise<StatusSummaryResponse> {
    const response = await apiClient.get<StatusSummaryResponse>(
      `${EVENTS_API_PATH}/${eventCode}/speakers/status-summary`
    );

    return response.data;
  }
}

// Export singleton instance
export const speakerStatusService = new SpeakerStatusService();

// Export class for testing
export default SpeakerStatusService;
